### PR TITLE
Back refactoring

### PR DESCRIPTION
**Pull request by marzieafshar**

### CODE REVIEW

**Review by M79Montazer — CHANGES_REQUESTED** [review]
**DO NOT MERGE THIS PR**
since we have a lot of changes in this PR, it needs to be tested properly before being merged with main. only after that, we can merge it with main.

**Review by msrazavi — APPROVED** [review]

**Review by aryajalali — APPROVED** [review]

**Review by M79Montazer — APPROVED** [review]

**Pull request merged**
